grpc-server/tinkerbell.go

```diff
@@ -2,47 +2,219 @@ package grpcserver
 func (s *server) GetWorkflowMetadata(context context.Context, req *pb.GetWorkflowDataRequest) (*pb.GetWorkflowDataResponse, error) {
-	return exec.GetWorkflowMetadata(context, req, s.db)
+	data, err := s.db.GetWorkflowMetadata(context, req)
+	if err != nil {
+		return &pb.GetWorkflowDataResponse{Data: []byte("")}, status.Errorf(codes.Unknown, err.Error())
+	}
+	return &pb.GetWorkflowDataResponse{Data: data}, nil
 }
 
 // GetWorkflowDataVersion returns the latest version of data for a workflow
 func (s *server) GetWorkflowDataVersion(context context.Context, req *pb.GetWorkflowDataRequest) (*pb.GetWorkflowDataResponse, error) {
-	return exec.GetWorkflowDataVersion(context, req.WorkflowID, s.db)
+	version, err := s.db.GetWorkflowDataVersion(context, req.WorkflowID)
+	if err != nil {
+		return &pb.GetWorkflowDataResponse{Version: version}, status.Errorf(codes.Unknown, err.Error())
+	}
+	return &pb.GetWorkflowDataResponse{Version: version}, nil
+}
+
+func getWorkflowsForWorker(db db.Database, id string) ([]string, error) {
+	if len(id) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "worker_id is invalid")
+	}
+	wfs, _ := db.GetWorkflowsForWorker(id)
+	if wfs == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Worker not found for any workflows")
+	}
+	return wfs, nil
+}
+
+func getWorkflowActions(context context.Context, db db.Database, wfID string) (*pb.WorkflowActionList, error) {
+	actions, err := db.GetWorkflowActions(context, wfID)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "workflow_id is invalid")
+	}
+	return actions, nil
+}
+
+// The below function check whether a particular workflow context is applicable or needed to
+// be send to a worker based on the state of the current action and the targeted workerID.
+func isApplicableToSend(context context.Context, wfContext *pb.WorkflowContext, workerID string, db db.Database) bool {
+	if wfContext.GetCurrentActionState() == pb.ActionState_ACTION_FAILED ||
+		wfContext.GetCurrentActionState() == pb.ActionState_ACTION_TIMEOUT {
+		return false
+	}
+	actions, err := getWorkflowActions(context, db, wfContext.GetWorkflowId())
+	if err != nil {
+		return false
+	}
+	if wfContext.GetCurrentActionState() == pb.ActionState_ACTION_SUCCESS {
+		if isLastAction(wfContext, actions) {
+			return false
+		}
+		if wfContext.GetCurrentActionIndex() == 0 {
+			if actions.ActionList[wfContext.GetCurrentActionIndex()+1].GetWorkerId() == workerID {
+				log.Println("Send the workflow context ", wfContext.GetWorkflowId())
+				return true
+			}
+		}
+	} else {
+		if actions.ActionList[wfContext.GetCurrentActionIndex()].GetWorkerId() == workerID {
+			log.Println("Send the workflow context ", wfContext.GetWorkflowId())
+			return true
+		}
+	}
+	return false
+}
+
+func isLastAction(wfContext *pb.WorkflowContext, actions *pb.WorkflowActionList) bool {
+	return int(wfContext.GetCurrentActionIndex()) == len(actions.GetActionList())-1
```

Comment:
return type of `GetCurrentActionIndex` is already int. So, int conversion won't be required?

Reply:
It is required because `GetCurrentActionIndex` returns an `int64` while `len` returns an `int`.